<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Li Lin" />

<meta name="date" content="2020-05-01" />

<title>ISnorm Package</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">ISnorm Package</h1>
<h4 class="author">Li Lin</h4>
<h4 class="date">2020-05-01</h4>



<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>ISnorm is a method implemented in R for normalizing single-cell RNA sequencing (scRNA-seq) data by a set of constantly expressed genes across all cells (internal spike-in genes, IS genes). We will demonstrate how to normalize scRNA-seq data using ISnorm in this tutorial.</p>
</div>
<div id="quick-start" class="section level1">
<h1>Quick Start</h1>
<div id="isnorm-input" class="section level2">
<h2>ISnorm input</h2>
<p>In this tutorial, we use UMI matrix data from <a href="https://linkinghub.elsevier.com/retrieve/pii/S0092867415005000">Klein et al</a>, which is available at GEO database under accession number <strong>GSE65525</strong> (GSM1599494_ES_d0_main.csv). But generally the inputs can be in many forms, including un-normalized matrix such as reads count and transcripts count, or normalized matrix such as rpm , tpm and fpkm (see our article for more details). The rows in <strong>mat</strong> object correspond to genes and columns correspond to cells.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ISnorm)
<span class="kw">library</span>(dbscan)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># check raw counts matrix</span>
mat[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##               cell1 cell2 cell3
## 0610005C13Rik     0     3     1
## 0610007P14Rik     1     3     2
## 0610009B22Rik     0     3     0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># In order to decrease run time, we randomly select 5000 genes and 300 cells</span>
<span class="kw">set.seed</span>(<span class="dv">1</span>)
sub.mat=mat[<span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(mat), <span class="dt">size =</span> <span class="dv">5000</span>),<span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(mat), <span class="dt">size =</span> <span class="dv">300</span>)]
<span class="kw">nrow</span>(sub.mat)</code></pre></div>
<pre><code>## [1] 5000</code></pre>
</div>
<div id="calculate-distance-between-genes" class="section level2">
<h2>Calculate distance between genes</h2>
<p>This step is the most time-consuming and running time mainly depends on the number of genes retained after filtering. You can try higher detection ratio <code>detection_rate=0.95</code> to filter more genes or use multiple cores <code>ncore=5</code> to reduce the running time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_dis&lt;-<span class="kw">calculate.dis</span>(sub.mat,<span class="dt">detection_rate=</span><span class="fl">0.90</span>)</code></pre></div>
<p><code>detection_rate=0.9</code> means genes without at least 90% cells having nonzero expression will not be included in further analyis.</p>
</div>
<div id="use-dbscan-algorithm-to-predict-is-genes" class="section level2">
<h2>Use DBscan algorithm to predict IS genes</h2>
<p>Next we use DBscan algorithm to predict IS genes. <code>ngene</code> specifies a series of expected number of IS genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spike_candidate&lt;-<span class="kw">dbscan.pick</span>(<span class="dt">dis=</span>gene_dis,<span class="dt">ngene=</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">floor</span>(<span class="kw">nrow</span>(gene_dis)<span class="op">/</span><span class="dv">25</span>))<span class="op">*</span><span class="dv">5</span>)</code></pre></div>
<p>After this step, which will return a list with each element containing one candidate geneset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(spike_candidate)</code></pre></div>
<pre><code>## List of 7
##  $ : chr [1:13] &quot;Eef1a1&quot; &quot;Rpl9&quot; &quot;Rps18&quot; &quot;Rpl36&quot; ...
##  $ : chr [1:17] &quot;Eef1a1&quot; &quot;Gm13826&quot; &quot;Rpl9&quot; &quot;Ppia&quot; ...
##  $ : chr [1:20] &quot;Eef1a1&quot; &quot;Gm13826&quot; &quot;Rpl9&quot; &quot;Ppia&quot; ...
##  $ : chr [1:25] &quot;Eef1a1&quot; &quot;Gm6654&quot; &quot;Gm13826&quot; &quot;Rpl9&quot; ...
##  $ : chr [1:31] &quot;Eef1a1&quot; &quot;Gm6654&quot; &quot;Gm13826&quot; &quot;Rpl9&quot; ...
##  $ : chr [1:35] &quot;Eef1a1&quot; &quot;Gm6654&quot; &quot;Hnrnpf&quot; &quot;Gm13826&quot; ...
##  $ : chr [1:46] &quot;Eef1a1&quot; &quot;Gm6654&quot; &quot;Hnrnpf&quot; &quot;Gm13826&quot; ...</code></pre>
</div>
<div id="normalize-matrix" class="section level2">
<h2>Normalize matrix</h2>
<p>In this step, We use each candidate set to normalize raw counts matrix. This function returns a list containing the normalization results for each candidate set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">candidate_res&lt;-<span class="kw">candidate.norm</span>(<span class="dt">mat=</span>sub.mat,<span class="dt">spike_candidate=</span>spike_candidate)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(candidate_res)</code></pre></div>
<pre><code>## [1] &quot;sf&quot;    &quot;inst&quot;  &quot;ngene&quot; &quot;ref&quot;   &quot;spike&quot;</code></pre>
<p><code>candidate_res$sf</code> is a numeric matrix containing the size factors, with rows representing cells and columns representing size factors estimated by each candidate set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">candidate_res<span class="op">$</span>sf[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##              [,1]      [,2]      [,3]
## cell248 1.0361437 1.0455692 1.0520635
## cell347 0.8900558 0.8503075 0.7934766
## cell534 1.6170666 1.5613605 1.5449976</code></pre>
<p><code>candidate_res$inst</code> is a numeric matrix containing the instability scores, with each row representing one cell and each column representing instability scores estimated by one candidate set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">candidate_res<span class="op">$</span>inst[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##               [,1]       [,2]       [,3]
## cell248 0.09946956 0.11323383 0.10845678
## cell347 0.07488380 0.07313330 0.08438882
## cell534 0.09269553 0.08528726 0.08261505</code></pre>
<p>you can draw a boxplot to see the instability score of cells for each candidate set to measure the reliability of each candidate set and manually choose best IS geneset (see our article for more details).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot every candidate geneset instability score</span>
candidate_res<span class="op">$</span>inst[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>##               [,1]       [,2]       [,3]
## cell248 0.09946956 0.11323383 0.10845678
## cell347 0.07488380 0.07313330 0.08438882
## cell534 0.09269553 0.08528726 0.08261505</code></pre>
</div>
<div id="choose-the-best-is-geneset" class="section level2">
<h2>Choose the best IS geneset</h2>
<p>You can manually choose best IS geneset based on the results from <code>candidate_res</code>. To be convenient, we have developed a method to automatically choose the best IS geneset based on <em>F-test</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ISnorm_res&lt;-<span class="kw">opt.candidate</span>(<span class="dt">mat=</span>mat,<span class="dt">candidate_res=</span>candidate_res)</code></pre></div>
<pre><code>## Candidate set 2 is chosen.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(ISnorm_res)</code></pre></div>
<pre><code>## [1] &quot;normalized&quot;  &quot;size_factor&quot; &quot;ISgenes&quot;     &quot;inst_cell&quot;   &quot;picked&quot;</code></pre>
<p><code>ISnorm_res$normalized</code> contains the normalized matrix. <code>ISnorm_res$size_factor</code> contains the size factor for each cell. <code>ISnorm_res$ISgenes</code> contains the name of IS genes used for normalization. <code>ISnorm_res$inst_cell</code> contains the instability score for each cell. <code>ISnorm_res$picked</code> contains the index of the optimized candidate geneset.</p>
<p>Lastly, you can use <code>ISnorm_res$normalized</code> and <code>ISnorm_res$size_factor</code> result to do downstream analysis for your single cell data.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
